numpy/lib/npyio.py

```diff
@@ -1051,23 +1051,27 @@ def read_data(chunk_size):
         byte_converters = True
 
     if usecols is not None:
-        # Allow usecols to be a single int or a sequence of ints
+        # Copy usecols, allowing it to be a single int or a sequence of ints.
         try:
-            usecols_as_list = list(usecols)
+            usecols = list(usecols)
         except TypeError:
-            usecols_as_list = [usecols]
-        for col_idx in usecols_as_list:
+            usecols = [usecols]
+        for i, col_idx in enumerate(usecols):
+            # Immediately cast to builtin ints; otherwise e.g. usecols[0] + 1
+            # below can become a float (because uint64 + 1 => float).
             try:
-                opindex(col_idx)
+                usecols[i] = opindex(col_idx)
             except TypeError as e:
                 e.args = (
                     "usecols must be an int or a sequence of ints but "
                     "it contains at least one element of type %s" %
                     type(col_idx),
                     )
                 raise
-        # Fall back to existing code
-        usecols = usecols_as_list
+        usecols_getter = (
+            itemgetter(*usecols) if len(usecols) > 1 else
+            # Get an iterable back, even if using a single column.
+            itemgetter(slice(usecols[0], usecols[0] + 1)))
```

Comment:
I think using `itemgetter` is overkill here - perhaps better as:
```suggestion
        # performance optimization - do not iterate `usecols` on every row if not needed
        if len(usecols) > 1:
            usecols_getter = lambda obj: [obj[c] for c in usecols]
        else:
            usecols_getter = lambda obj, c=usecols[0]: [obj[c]]
```

Reply:
For the len(usecols)>1 case itemgetter is clearly faster by ~10% (avoiding Python-level iteration (`for c in usecols`) is probably where the speedup comes from); for the length-1 case there's no difference in speed but I guess the lambda is indeed clearer.

Reply:
`itemgetter` is implemented in Python as far as I can tell, so I can't see why it would be faster.

Reply:
It isn't, see https://github.com/python/cpython/blob/6bd3ecfc272b122b55a6adec50dd7a7c868f262f/Modules/_operator.c#L6 (but you may have been confused because https://github.com/python/cpython/blob/7dad0337518a0d599caf8f803a5bf45db67cbe9b/Lib/operator.py#L408 overwrites https://github.com/python/cpython/blob/7dad0337518a0d599caf8f803a5bf45db67cbe9b/Lib/operator.py#L265?)

Reply:
Ah, I was actually confused because `inspect.getsource` is lying to me!

Reply:
yes, I got bitten by that too...